File: .agents/skills/open-knowledge-format/scripts/migrate_v01_to_v02.rb
#!/usr/bin/env ruby

require 'date'
require 'json'
require 'pathname'
require 'time'
require 'yaml'

root = Pathname(ARGV.fetch(0)).expand_path
actor = ARGV.fetch(1)
generated_at = ARGV.fetch(2, Time.now.utc.iso8601)
actor_pattern = %r{\A(?:[^/:\s]+/[^\s]+|human:[^\s]+|process:[^\s]+)\z}

abort "bundle root does not exist: #{root}" unless root.directory?
abort 'actor must use producer/version, human:<id>, or process:<id>' unless actor.match?(actor_pattern)
begin
  raise ArgumentError unless generated_at.match?(/\A\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})\z/)
  Time.iso8601(generated_at)
rescue ArgumentError
  abort 'generated_at must be an ISO 8601 datetime'
end

def yaml_string(value)
  JSON.generate(value)
end

def parse_frontmatter(path, text)
  match = text.match(/\A---[ \t]*\r?\n(.*?)\r?\n---[ \t]*\r?\n/m)
  abort "missing frontmatter: #{path}" unless match
  begin
    data = YAML.safe_load(match[1], permitted_classes: [Date, Time], aliases: false) || {}
  rescue StandardError => error
    abort "invalid frontmatter in #{path}: #{error.message}"
  end
  abort "frontmatter must be a mapping: #{path}" unless data.is_a?(Hash)
  [match, data]
end

def sources_from(citations)
  links = []
  markdown_link = /\[([^\]]+)\]\(((?:[^()\s]+|\([^()]*\))+|<[^>]+>)\)/
  residual = citations.gsub(markdown_link) do
    title = Regexp.last_match(1).strip
    resource = Regexp.last_match(2).delete_prefix('<').delete_suffix('>')
    links << [title, resource]
    ''
  end
  residual.scan(%r{https?://[^\s<>)]+}).each { |resource| links << [resource, resource] }

  links.uniq.each_with_index.map do |(title, resource), index|
    { 'id' => "citation-#{index + 1}", 'resource' => resource, 'title' => title }
  end
end

concepts = root.glob('**/*.md').reject { |path| %w[index.md log.md].include?(path.basename.to_s) }
migrated = 0
concepts.each do |path|
  text = path.read(encoding: 'UTF-8')
  match, data = parse_frontmatter(path, text)
  frontmatter = match[1]
  body = text[match.end(0)..]
  citation_match = body.match(/(?:\A|\r?\n)# Citations[ \t]*\r?\n+(.*)\z/m)
  has_legacy_metadata = data.key?('timestamp') || citation_match

  next if data.key?('generated') && !has_legacy_metadata
  abort "mixed v0.1 and v0.2 generation metadata: #{path}" if data.key?('generated')
  abort "cannot merge an existing sources field with legacy citations: #{path}" if data.key?('sources') && citation_match

  sources = citation_match ? sources_from(citation_match[1]) : []
  abort "legacy citation section contains no extractable resources: #{path}" if citation_match && sources.empty? && !citation_match[1].strip.empty?
  body = body[0...citation_match.begin(0)].rstrip + "\n" if citation_match

  frontmatter = frontmatter.lines.reject { |line| line.match?(/\Atimestamp:[ \t]*/) }.join.rstrip
  unless sources.empty?
    frontmatter += "\nsources:\n"
    sources.each do |source|
      frontmatter += "  - id: #{yaml_string(source.fetch('id'))}\n"
      frontmatter += "    resource: #{yaml_string(source.fetch('resource'))}\n"
      frontmatter += "    title: #{yaml_string(source.fetch('title'))}\n"
    end
  end
  frontmatter += "\ngenerated:\n  by: #{yaml_string(actor)}\n  at: #{yaml_string(generated_at)}"

  path.write("---\n#{frontmatter}\n---\n\n#{body.lstrip}")
  migrated += 1
end

root_index = root.join('index.md')
if root_index.exist?
  text = root_index.read(encoding: 'UTF-8')
  if text.start_with?('---')
    match, data = parse_frontmatter(root_index, text)
    abort "root index frontmatter may contain only okf_version: #{root_index}" unless data.keys == ['okf_version']
    root_index.write(text[0...match.begin(0)] + "---\nokf_version: \"0.2\"\n---\n" + text[match.end(0)..])
  else
    root_index.write("---\nokf_version: \"0.2\"\n---\n\n#{text}")
  end
end

root.glob('**/log.md').each do |path|
  text = path.read(encoding: 'UTF-8')
  text = "# Knowledge Bundle Update Log\n\n#{text}" if text.match?(/\A# \d{4}-\d{2}-\d{2}[ \t]*$/)
  text = text.gsub(/^# (\d{4}-\d{2}-\d{2})[ \t]*$/, '## \1')
  path.write(text)
end

puts "Migrated #{migrated} concepts in #{root} to the OKF v0.2 producer profile."

File: .agents/skills/open-knowledge-format/scripts/migrate-v01-to-v02.mjs
#!/usr/bin/env node

import { readdir, readFile, stat, writeFile } from 'node:fs/promises';
import path from 'node:path';
import { pathToFileURL } from 'node:url';

import yaml from 'js-yaml';

const actorPattern = /^(?:[^/:\s]+\/[^\s]+|human:[^\s]+|process:[^\s]+)$/u;
const datetimePattern = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})$/u;

export async function migrateV01ToV02(bundleRoot, actor, generatedAt = new Date().toISOString().replace('.000Z', 'Z')) {
  const root = path.resolve(bundleRoot);
  if (!(await isDirectory(root))) throw new Error(`bundle root does not exist: ${root}`);
  if (!actorPattern.test(actor)) throw new Error('actor must use producer/version, human:<id>, or process:<id>');
  if (!datetimePattern.test(generatedAt) || Number.isNaN(Date.parse(generatedAt))) {
    throw new Error('generated_at must be an ISO 8601 datetime');
  }

  const concepts = (await markdownFiles(root)).filter(
    (filePath) => !['index.md', 'log.md'].includes(path.basename(filePath)),
  );
  let migrated = 0;
  for (const filePath of concepts) {
    const text = await readFile(filePath, 'utf8');
    const parsed = parseFrontmatter(filePath, text);
    let { frontmatter } = parsed;
    let body = parsed.body;
    const citationMatch = /(?:^|\r?\n)# Citations[ \t]*\r?\n+(.*)$/su.exec(body);
    const hasLegacyMetadata = Object.hasOwn(parsed.data, 'timestamp') || citationMatch !== null;

    if (Object.hasOwn(parsed.data, 'generated') && !hasLegacyMetadata) continue;
    if (Object.hasOwn(parsed.data, 'generated')) {
      throw new Error(`mixed v0.1 and v0.2 generation metadata: ${filePath}`);
    }
    if (Object.hasOwn(parsed.data, 'sources') && citationMatch !== null) {
      throw new Error(`cannot merge an existing sources field with legacy citations: ${filePath}`);
    }

    const sources = citationMatch === null ? [] : sourcesFrom(citationMatch[1]);
    if (citationMatch !== null && sources.length === 0 && citationMatch[1].trim().length > 0) {
      throw new Error(`legacy citation section contains no extractable resources: ${filePath}`);
    }
    if (citationMatch !== null) body = `${body.slice(0, citationMatch.index).trimEnd()}\n`;

    frontmatter = frontmatter
      .split(/(?<=\n)/u)
      .filter((line) => !/^timestamp:[ \t]*/u.test(line))
      .join('')
      .trimEnd();
    if (sources.length > 0) {
      frontmatter += '\nsources:\n';
      for (const source of sources) {
        frontmatter += `  - id: ${JSON.stringify(source.id)}\n`;
        frontmatter += `    resource: ${JSON.stringify(source.resource)}\n`;
        frontmatter += `    title: ${JSON.stringify(source.title)}\n`;
      }
    }
    frontmatter += `\ngenerated:\n  by: ${JSON.stringify(actor)}\n  at: ${JSON.stringify(generatedAt)}`;

    await writeFile(filePath, `---\n${frontmatter}\n---\n\n${body.trimStart()}`);
    migrated += 1;
  }

  const rootIndex = path.join(root, 'index.md');
  if (await isFile(rootIndex)) {
    const text = await readFile(rootIndex, 'utf8');
    if (text.startsWith('---')) {
      const parsed = parseFrontmatter(rootIndex, text);
      const keys = Object.keys(parsed.data);
      if (keys.length !== 1 || keys[0] !== 'okf_version') {
        throw new Error(`root index frontmatter may contain only okf_version: ${rootIndex}`);
      }
      await writeFile(rootIndex, `---\nokf_version: "0.2"\n---\n${parsed.body}`);
    } else {
      await writeFile(rootIndex, `---\nokf_version: "0.2"\n---\n\n${text}`);
    }
  }

  for (const filePath of (await markdownFiles(root)).filter((candidate) => path.basename(candidate) === 'log.md')) {
    let text = await readFile(filePath, 'utf8');
    if (/^# \d{4}-\d{2}-\d{2}[ \t]*$/u.test(text.split(/\r?\n/u)[0] ?? '')) {
      text = `# Knowledge Bundle Update Log\n\n${text}`;
    }
    text = text.replace(/^# (\d{4}-\d{2}-\d{2})[ \t]*$/gmu, '## $1');
    await writeFile(filePath, text);
  }

  return migrated;
}

function parseFrontmatter(filePath, text) {
  const match = /^---[ \t]*\r?\n(.*?)\r?\n---[ \t]*\r?\n/su.exec(text);
  if (match === null) throw new Error(`missing frontmatter: ${filePath}`);
  let data;
  try {
    data = yaml.load(match[1], { schema: yaml.JSON_SCHEMA }) ?? {};
  } catch (error) {
    throw new Error(`invalid frontmatter in ${filePath}: ${error instanceof Error ? error.message : String(error)}`);
  }
  if (typeof data !== 'object' || data === null || Array.isArray(data)) {
    throw new Error(`frontmatter must be a mapping: ${filePath}`);
  }
  return { body: text.slice(match[0].length), data, frontmatter: match[1] };
}

function sourcesFrom(citations) {
  const links = [];
  const markdownLink = /\[([^\]]+)\]\(((?:[^()\s]+|\([^()]*\))+|<[^>]+>)\)/gu;
  const residual = citations.replace(markdownLink, (_match, rawTitle, rawResource) => {
    links.push([rawTitle.trim(), rawResource.replace(/^</u, '').replace(/>$/u, '')]);
    return '';
  });
  for (const match of residual.matchAll(/https?:\/\/[^\s<>)]+/gu)) links.push([match[0], match[0]]);

  const seen = new Set();
  const unique = [];
  for (const [title, resource] of links) {
    const key = `${title}\0${resource}`;
    if (seen.has(key)) continue;
    seen.add(key);
    unique.push({ id: `citation-${unique.length + 1}`, resource, title });
  }
  return unique;
}

async function markdownFiles(root) {
  const files = [];
  await visit(root, files);
  return files.sort();
}

async function visit(directory, files) {
  for (const entry of await readdir(directory, { withFileTypes: true })) {
    const absolutePath = path.join(directory, entry.name);
    if (entry.isDirectory()) await visit(absolutePath, files);
    else if (entry.isFile() && entry.name.endsWith('.md')) files.push(absolutePath);
  }
}

async function isDirectory(filePath) {
  try {
    return (await stat(filePath)).isDirectory();
  } catch (error) {
    if (error?.code === 'ENOENT') return false;
    throw error;
  }
}

async function isFile(filePath) {
  try {
    return (await stat(filePath)).isFile();
  } catch (error) {
    if (error?.code === 'ENOENT') return false;
    throw error;
  }
}

if (isMainModule(import.meta.url)) {
  try {
    const root = process.argv[2];
    const actor = process.argv[3];
    if (root === undefined || actor === undefined) {
      throw new Error('usage: migrate-v01-to-v02.mjs <bundle-root> <actor> [generated-at]');
    }
    const migrated = await migrateV01ToV02(root, actor, process.argv[4]);
    process.stdout.write(`Migrated ${migrated} concepts in ${path.resolve(root)} to the OKF v0.2 producer profile.\n`);
  } catch (error) {
    process.stderr.write(`${error instanceof Error ? error.message : String(error)}\n`);
    process.exitCode = 1;
  }
}

function isMainModule(url) {
  return process.argv[1] !== undefined && url === pathToFileURL(path.resolve(process.argv[1])).href;
}
